twenty-first/src/math/digest.rs

```diff
@@ -145,22 +153,26 @@ impl From<Digest> for [u8; Digest::BYTES] {
     }
 }
 
-impl From<[u8; Digest::BYTES]> for Digest {
-    fn from(item: [u8; Digest::BYTES]) -> Self {
-        let chunk_into_bfe = |chunk: &[u8]| {
+impl TryFrom<[u8; Digest::BYTES]> for Digest {
+    type Error = TryFromDigestError;
+
+    fn try_from(item: [u8; Digest::BYTES]) -> Result<Self, Self::Error> {
+        fn chunk_into_bfe(chunk: &[u8]) -> Result<BFieldElement, TryFromDigestError> {
             let mut arr = [0u8; BFieldElement::BYTES];
             arr.copy_from_slice(chunk);
-            BFieldElement::from(arr)
-        };
+            let int = u64::from_le_bytes(arr);
+            match BFieldElement::is_canonical(int) {
+                true => Ok(BFieldElement::from(arr)),
+                false => Err(TryFromDigestError::NotCanonical(int)),
+            }
```

Comment:
I think I prefer this over matching on a `bool`[^1].

```suggestion
            BFieldElement::is_canonical(int)
                .then(|| BFieldElement::new(int))
                .ok_or(TryFromDigestError::NotCanonical(int))
```

[^1]: I have used such matches in the past but dislike their readability now.

Reply:
ok, will do.